hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CreatePipelineCommandHandler.java

```diff
@@ -0,0 +1,228 @@
+            .setCreatePipelineAck(createPipelineACK);
+      };
+      updateCommandStatus(context, command, statusUpdater, LOG);
+      long endTime = Time.monotonicNow();
+      totalTime += endTime - startTime;
+    }
+  }
+
+  /**
+   * Returns the command type that this command handler handles.
+   *
+   * @return Type
+   */
+  @Override
+  public SCMCommandProto.Type getCommandType() {
+    return SCMCommandProto.Type.createPipelineCommand;
+  }
+
+  /**
+   * Returns number of times this handler has been invoked.
+   *
+   * @return int
+   */
+  @Override
+  public int getInvocationCount() {
+    return (int)invocationCount.get();
+  }
+
+  /**
+   * Returns the average time this function takes to run.
+   *
+   * @return long
+   */
+  @Override
+  public long getAverageRunTime() {
+    if (invocationCount.get() > 0) {
+      return totalTime / invocationCount.get();
+    }
+    return 0;
+  }
+
+  /**
+   * Sends ratis command to create pipeline on this datanode.
+   *
+   * @param dn       - this datanode
+   * @param pipelineId   - pipeline ID
+   * @param peers        - datanodes of the pipeline
+   * @param context      - Ozone datanode context
+   * @throws IOException
+   */
+  private void createPipeline(DatanodeDetails dn, PipelineID pipelineId,
+      Collection<DatanodeDetails> peers, StateContext context)
+      throws IOException {
+    final Configuration ozoneConf = context.getParent().getConf();
+    final RaftGroup group = RatisHelper.newRaftGroup(
+        RaftGroupId.valueOf(pipelineId.getId()), peers);
+    LOG.debug("creating pipeline:#{} with {}", pipelineId, group);
+
+    final String rpcType = ozoneConf
+        .get(ScmConfigKeys.DFS_CONTAINER_RATIS_RPC_TYPE_KEY,
+            ScmConfigKeys.DFS_CONTAINER_RATIS_RPC_TYPE_DEFAULT);
+    final RetryPolicy retryPolicy = RatisHelper.createRetryPolicy(ozoneConf);
+    final List< IOException > exceptions =
+        Collections.synchronizedList(new ArrayList<>());
+    final int maxOutstandingRequests =
+        HddsClientUtils.getMaxOutstandingRequests(ozoneConf);
+    final CertificateClient dnCertClient =
+        context.getParent().getCertificateClient();
+    final GrpcTlsConfig tlsConfig = RatisHelper.createTlsServerConfigForDN(
+        new SecurityConfig(ozoneConf), dnCertClient);
+    final TimeDuration requestTimeout =
+        RatisHelper.getClientRequestTimeout(ozoneConf);
+    try {
+      final RaftPeer peer = RatisHelper.toRaftPeer(dn);
+      try (RaftClient client = RatisHelper
+          .newRaftClient(SupportedRpcType.valueOfIgnoreCase(rpcType), peer,
+              retryPolicy, maxOutstandingRequests, tlsConfig,
+              requestTimeout)) {
+
+        RaftClientReply reply = client.groupAdd(group, peer.getId());
+        if (reply == null || !reply.isSuccess()) {
+          String msg = "Pipeline initialization failed for pipeline:"
+              + pipelineId.getId() + " node:" + peer.getId();
+          throw new IOException(msg);
+        }
```

Comment:
One question here: Is it even possible that we send out a command from SCM. The node picks it up and then tries to create the pipeline. It works, but the SCM command times out and SCM tries to create a pipeline again with those 3 data nodes.  This next command will fail, since these data nodes already have a pipeline. Should we try to handle that case ? On the SCM side, we will get an Incremental data node report, which will indicate that new pipeline is created. But the command might be marked as failed for the second retry. Should we try to make this create idempotent? If the pipeline exist, we are probably ok. @nandakumar131, @xiaoyuyao

Reply:
This might be an issue the current code as well -- just flagging for consideration; not really a comment on current patch.

Reply:
We can do this same idempotent code for deletePipeline and closePipeline as well ? @nandakumar131 , @xiaoyuyao

Reply:
When using async pipeline creation,  SCM send out the command to every datanode of a pipeline.  After the command is sent, create pipeline API will return as succeed.  Pipeline is in ALLOCATED state.  Once the SCM received all 3 datanodes's confirmation of successful pipeline allocation, SCM will change pipeline state from ALLOCATED to OPEN.  Pipeline is ready to servie the write request when it's state is OPEN.   So the possible scenerios is
1.  what if one of the datanode report failure 
2.  what if ond of the datanode's create status packet is lost
3.  what if one of the datanode dead after receive the command
For case 1,  current sync logic is as long as one datanode create pipeline fails(throw exception), this rewritten asyn create pipeline follow the same behavior. 
For case 2,  As long as the datanode is live,  through the status confirmation is lost, SCM can wait for datanode's pipeline report(default 60s interval ),  once pipeline report pipeline create success, SCM still can change the pipeline from ALLOCATED to OPEN state. 
For case 3,  SCM need a scrubber logic to detect all pipelines in ALLOCATD state,  and decide which one is timeout for a long time, and then close it, release the datanodes with this pipeline. 
Created HDDS-2177, Add a srubber thread to detect creation failure pipelines in ALLOCATED state to track this task. 

And agree, we can make closePipeline an idempotent operation.